### ISSUE TITLE
urlserver.py: incorrectly includes non-percent-encoded characters in URIs

### ISSUE THREAD

**Issue by MacGyverNL** [weechat/scripts#188]
Per [RFC3986, section 2](https://tools.ietf.org/html/rfc3986#section-2), any character not in either the reserved set or the unreserved set in a URI *must* be percent-encoded. Characters for which is the case are e.g. the space (%20), but also " (%22), < (%3C), > (%3E), { (%7B), and } (%7D). As it stands, urlserver.py seems to limit the end-match character class to the characters ], ), >, and the whitespace class \s. This is both incomplete and wrong: incomplete because it ignores such characters as " and <, and wrong because ) is a valid sub-delim for any URI scheme and thus cuts off many valid URLs.

This leads to annoying situations where e.g. shortened embedded tweet URIs from twitter's URI title scheme include the " in the redirect (twitter adds "s around the entire tweet in the title text of the page, so any tweet ending in a URI gets a wrong short link). Since title-pasting bots are rather common, so does this problem seem to be.

Another annoying problem is that valid wikipedia links are often cut before the end, since a lot of them include parentheses (), e.g. any page with a disambiguation (<https://en.wikipedia.org/wiki/Git_(disambiguation)>, and <https://en.wikipedia.org/wiki/Git_(slang)>).

As far as I can tell, [RFC3987](https://tools.ietf.org/html/rfc3987) (internationalized resource identifiers) does not change this situation.

My first thought to fix this is to change the matching regex to a positive rather than a negative match, and simply positive-match all the characters from the unreserved and reserved sets. However, this will probably not work for internationalized URLs. Rather, then, a negative-match on *all* the ASCII-characters not in the reserved and unreserved sets should work.

This may lead to the undesirable characteristic where people delimiting URLs with parentheses find their closing parenthesis matched by the regular expression. I'll state once again that they're doing it wrong, but since it is such a natural thing to do, maybe another match should be performed and if there exists a matching opening parenthesis before the URI scheme specifier for a closing parenthesis at the end of the URI, the closing parenthesis should be dropped from the match. Another option, that the URXVT terminal matcher.pl seems to prefer, is to include the closing parenthesis as part of the URI if and only if there is a matching opening parenthesis present anywhere in the URI. This could be limited to the part after the domain specifier, and could be part of a single negative-matching regex.

**Comment by MacGyverNL** [comment]
> My first thought to fix this is to change the matching regex to a positive rather than a negative match, and simply positive-match all the characters from the unreserved and reserved sets. However, this will probably not work for internationalized URLs.

Coming back to this, the class \w seems to match unicode encoded in UTF-8 just fine in my interactive python shell, so this actually becomes a viable option. The matcher.pl regular expression (listed below) seems to follow this approach (to some extent); a variation of this could be adopted.

However, I distilled the following from RFC3986/7, available in 315a2363485b7a8cd71a42ef39cadf4a047d07a7

```
url_scheme = r'[a-zA-Z][a-zA-Z0-9+\-.]*'

url_octet = r'(?:2(?:[0-4]\d|5[0-5])|1\d\d|\d{1,2})'
url_ipaddr = r'%s(?:\.%s){3}' % (url_octet, url_octet)

url_hexdig = r'[0-9a-fA-F]'
url_h16 = r'%s{1,4}' % (url_hexdig)
url_ls32 = r'%s:%s|%s' % (url_h16, url_h16, url_ipaddr)
url_ip6addr = [r'(?:%s:){6}%s' % (url_h16, url_ls32),
               r'::(?:%s:){5}%s' % (url_h16, url_ls32),
               r'%s?::(?:%s:){4}%s' % (url_h16, url_h16, url_ls32),
               r'(?:(?:%s:){0,1}%s)?::(?:%s:){3}%s' % (url_h16, url_h16, url_h16, url_ls32),
               r'(?:(?:%s:){0,2}%s)?::(?:%s:){2}%s' % (url_h16, url_h16, url_h16, url_ls32),
               r'(?:(?:%s:){0,3}%s)?::(?:%s:)%s' % (url_h16, url_h16, url_h16, url_ls32),
               r'(?:(?:%s:){0,4}%s)?::%s' % (url_h16, url_h16, url_ls32),
               r'(?:(?:%s:){0,5}%s)?::%s' % (url_h16, url_h16, url_h16),
               r'(?:(?:%s:){0,6}%s)?::' % (url_h16, url_h16)]
url_ip6addr = "(?:" + ")|(?:".join(url_ip6addr) + ")"
url_iplit = r'\[(?:%s)\]' % (url_ip6addr) # We're ignoring the IPvFuture

url_gendelims = r'[:/?#\[\]@]'
url_subdelims = r"[!$&'()*+,;=]"
url_reserved = r'(?:%s|%s)' % (url_gendelims, url_subdelims)
url_iunreserved = r'[\w\-.~]'

url_pctencoded = r'%%%s{2}' % (url_hexdig)

url_iregname = r'(?:%s|%s|%s)*' % (url_iunreserved, url_pctencoded, url_subdelims)

url_iuserinfo = r'(?:%s|%s|%s|:)*' % (url_iunreserved, url_pctencoded, url_subdelims)
url_ihost = r'(?:%s|%s|%s)' % (url_iplit, url_ipaddr, url_iregname)
url_iauth = r'(?:%s@)?%s(?::\d*)?' % (url_iuserinfo, url_ihost)

url_ipchar = r'(?:%s|%s|%s|:|@)' % (url_iunreserved, url_pctencoded,
                                    url_subdelims)
url_ipath_abempty = r'(?:/%s*)*' % (url_ipchar)

# Some complex stuff about reserved parts of the UCS namespace we're not doing
# in iquery, so iquery == ifragment.
url_iquery = r'(?:%s|/|\?)*' % (url_ipchar)
url_ifragment = url_iquery

# Grab one additional character (if present) so that we can later determine
# whether the user knew what they were doing.
#url_full = r'((?:%s)://(?:%s)(?:%s)(?:\?%s)?(?:#%s)?)(.)?' % (url_scheme, url_iauth,
url_full = r'(?:%s)://(?:%s)(?:%s)(?:\?%s)?(?:#%s)?' % (url_scheme, url_iauth,
                                                        url_ipath_abempty,
                                                        url_iquery,
                                                        url_ifragment)
```

with

`'regex': re.compile(url_full, re.IGNORECASE),`

I don't handle trailing closing parentheses here, that will need more thought and discussion about what a good heuristic might be, but I'm fairly certain that this provides complete and correct URI matching. Whether there are false positives / negatives will require testing.

**Comment by MacGyverNL** [comment]
Resolved (ages ago, whoops) by https://github.com/weechat/scripts/commit/c171586b75cff97518d1c902b879eb119e59e061
